### PR TITLE
Add featurizer evaluation subpackage (AKSW-13145)

### PR DESCRIPTION
**Pull request by rahulthakurakridata**
Added the following:
1. Evaluate functionality to compare different image featurizers (TF-HUB ones and also the one present in this repository).
2. Provides training and test accuracy with respect to different number of feature dimensions(using PCA).

### CODE REVIEW

File: DE-image-featurizer/evaluate/evaluator.py [hunk 1]
```diff
@@ -0,0 +1,375 @@
...
+class ImageFeatureEvaluator:
+    """
+    Provides the functionality to compare different image featurizers using KNN accuracy scores.
+    1. The dataset is divided into train and test sets.
+    2. Features are computed using the featurizer provided.
+    3. Comparative analysis is shown on different numbers of reduced dimensions using PCA.
+
+    Parameters
+    ----------
+
```
Comment (sabarish-akridata) [inline DE-image-featurizer/evaluate/evaluator.py:48]: Remove empty line after parameters.

File: DE-image-featurizer/evaluate/evaluator.py [hunk 2]
```diff
@@ -0,0 +1,375 @@
...
+        .........b_img1
+        .........b_img2
+        ```
+
+    Let's try on a simple tensorflow dataset.
+    >>> dataset_url = "https://storage.googleapis.com/download.tensorflow.org/example_images/flower_photos.tgz"
+    >>> archive = tf.keras.utils.get_file(origin=dataset_url, extract=True)
+    ...
+    >>> data_dir = pathlib.Path(archive).with_suffix('')
+    >>> fan = ImageFeatureEvaluator(data_dir,'https://tfhub.dev/google/imagenet/mobilenet_v1_025_224/feature_vector/5',32,False)
```
Comment (sabarish-akridata) [inline DE-image-featurizer/evaluate/evaluator.py]: Create a separate variable for `MODEL_URL` instead of making this line so long.

Comment (sabarish-akridata) [reply DE-image-featurizer/evaluate/evaluator.py]: Use a different name than `fan` for the object. Can't relate `fan` to what we are doing here. Call it `evaluator` instead.

File: DE-image-featurizer/evaluate/evaluator.py [hunk 3]
```diff
@@ -0,0 +1,375 @@
...
+    1. The dataset is divided into train and test sets.
+    2. Features are computed using the featurizer provided.
+    3. Comparative analysis is shown on different numbers of reduced dimensions using PCA.
+
+    Parameters
+    ----------
+
+    data_dir : string
+        Provides the path to parent directory where data is present
+    featurizer_selected : string
```
Comment (sabarish-akridata) [inline DE-image-featurizer/evaluate/evaluator.py]: rename to `featurizer`. The `_selected` suffix makes the variable name long without adding much info.
Add some extra info like so:
```
featurizer: string
    Name of a traditional CV featurizer, or URL to a tf-hub featurizer model.
    Accepted values for names are 'hog', 'lbp', 'orb', 'sift', <include other names>
```

File: DE-image-featurizer/evaluate/evaluator.py [hunk 4]
```diff
@@ -0,0 +1,375 @@
...
+        Hyperlink to the Image featurizer module
+    batch_size: integer
+        Size of batches of data.
+    structure: boolean
```
Comment (sabarish-akridata) [inline DE-image-featurizer/evaluate/evaluator.py]: Rename this to `splits`.

File: DE-image-featurizer/evaluate/evaluator.py [hunk 5]
```diff
@@ -0,0 +1,375 @@
...
+    Let's try on a simple tensorflow dataset.
+    >>> dataset_url = "https://storage.googleapis.com/download.tensorflow.org/example_images/flower_photos.tgz"
+    >>> archive = tf.keras.utils.get_file(origin=dataset_url, extract=True)
+    ...
+    >>> data_dir = pathlib.Path(archive).with_suffix('')
+    >>> fan = ImageFeatureEvaluator(data_dir,'https://tfhub.dev/google/imagenet/mobilenet_v1_025_224/feature_vector/5',32,False)
+
+    >>> train_ds, _ = fan.load_data()
+    Found ...
+    >>> featurizer, preprocessor = fan.load_featurizer()
```
Comment (sabarish-akridata) [inline DE-image-featurizer/evaluate/evaluator.py]: No need to return these and pass them to the next method. Assign these as attributes of the object inside the `load_featurizer` method, and use them in the `extract_features` method.

File: DE-image-featurizer/evaluate/evaluator.py [hunk 6]
```diff
@@ -0,0 +1,380 @@
+"""
+evaluate.py
+Provides the functinality to compare different image featurizers
+Creates a featurizer from tf-hub or traditional repositoy based models.
+Applies PCA to reduce feature dimensions.
+Provides KNN accuracy scores for different set of dimensions
+"""
+import sys
```
Comment (raghav-akridata) [inline DE-image-featurizer/evaluate/evaluator.py:9]: @rahulthakurakridata Please organise and optimise imports. Refer [here](https://www.jetbrains.com/pycharm/guide/tips/optimize-imports/).

File: DE-image-featurizer/evaluate/evaluator.py [hunk 7]
```diff
@@ -0,0 +1,380 @@
...
+from sklearn.metrics import accuracy_score
+from sklearn.decomposition import PCA
+import pandas as pd
+import tensorflow_hub as hub
+sys.path.append("")
+from featurize.descriptor import HOG, LBP, ORB, SIFT
+
+
+
+DEF_N_COMPONENTS = [2, 8, 16, 32, 64, 128]  # Different number of components used during PCA.
```
Comment (raghav-akridata) [inline DE-image-featurizer/evaluate/evaluator.py]: @rahulthakurakridata Let's follow an uniform comment structure. Comments followed by a code line.

Something like this:
```python
# Different number of components used during PCA.
DEF_N_COMPONENTS = [2, 8, 16, 32, 64, 128]
# The dictionary describing different traditional CV featurizer.
FEATURIZER_DICT = {'hog': HOG(), 'lbp': LBP(), 'orb': ORB(), 'sift': SIFT()}
# Random seed for shuffling and transformations
SEED = 123  
```

Comment (raghav-akridata) [reply DE-image-featurizer/evaluate/evaluator.py]: For very short comments 2-3 words we can still write inline comments. The judgment still lies with you. But keep in mind uniformity increases readability.

File: DE-image-featurizer/evaluate/evaluator.py [hunk 8]
```diff
@@ -0,0 +1,380 @@
...
+            Numpy array containing corresponding labels
+
+        """
+        features_array = None
+        input_labels = None
+        input_data = self.train_ds if is_train else self.val_ds
+        for image_batch, labels_batch in tqdm(input_data):
+            preprocessed_image_batch = self.preprocessor(image_batch)
+            image_batch_features = self.featurizer(preprocessed_image_batch)
+            if features_array is None:
```
Comment (raghav-akridata) [inline DE-image-featurizer/evaluate/evaluator.py]: @rahulthakurakridata Try avoiding multiple if-else statements. Maybe, you can append all features to a list and do `np.asarray` over it.

**Review by raghav-akridata — CHANGES_REQUESTED** [review]

**Review by raghav-akridata — APPROVED** [review]
